File: src/main/scala/Interview/scalaCode/zzke/B3_2.scala
package Interview.scalaCode.zzke

import java.util.Date

class Matrix(private var n: Long) {
  val mod = 1000000007
  var m = n-1

  def ope(a: (Long, Long), b: (Long, Long)): (Long, Long) = ((a._1 * b._1 + a._2 * b._2) % mod, (a._1 * b._2 + a._2 * b._1 + a._2 * b._2) % mod)

  def res: Long = {
    var a: (Long, Long) = (1, 0)
    var b: (Long, Long) = (0, 1)

    while (m != 0) {
      if (m % 2 == 1) a = ope(a, b)
      b = ope(b, b)
      m = m >> 1
    }
    (a._1 + a._2) % mod

  }

}

object B3_2 extends App {

  val n = 1000000000
  val a1: Long = new Matrix(n).res
  val start = new Date().getTime
  println(a1)
  val end = new Date().getTime
  println(end - start)

}

File: src/main/scala/Interview/javaCode/zzke/B3_2.java
package Interview.javaCode.zzke;

/**
 * @Author: chenjinxin
 * @Date: Created in 下午4:00 19-9-29
 * @Description:
 */

public class B3_2 {
    final static long mod = 1000000007;

    public static Pair make_pair(long a, long b) {
        return new Pair(a, b);
    }

    public static Pair ope(Pair a, Pair b) {
        return make_pair((a.first * b.first + a.second * b.second) % mod, (a.first * b.second + a.second * b.first + a.second * b.second) % mod);
    }

    public static long F(long y) {
        Pair ans = make_pair(1, 0);
        Pair x = make_pair(0, 1);
        while (y != 0) {
            if (y % 2 == 1) ans = ope(ans, x);
            x = ope(x, x);
            y = y >> 1;
        }
        return ans.first + ans.second;
    }

    public static void main(String[] args) {
        long t1 = System.currentTimeMillis();
        long n = 1000000000;
        long result = 0;
        result = F(n - 1) % mod;
        System.out.println(result);
        long t2 = System.currentTimeMillis();
        System.out.println(t2 - t1);
    }
}

class Pair {
    long first;
    long second;

    public Pair(long first, long second) {
        this.first = first;
        this.second = second;
    }

    public long getFirst() {
        return first;
    }

    public void setFirst(long first) {
        this.first = first;
    }

    public long getSecond() {
        return second;
    }

    public void setSecond(int second) {
        this.second = second;
    }
}
